pkg/cloud/ovirt/providerIDcontroller/providerIDController.go

```diff
@@ -0,0 +1,150 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.log.Error(err, "Failed getting node")
+		return reconcile.Result{}, fmt.Errorf("error getting node: %v", err)
+	}
+
+	if node.Spec.ProviderID != "" {
+		return reconcile.Result{}, nil
+	}
+
+	r.log.Info("spec.ProviderID is empty, fetching from ovirt")
+	id, err := r.fetchProviderIDFunc(node.Name)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed getting VM from oVirt: %v", err)
+	}
+
+	node.Spec.ProviderID = fmt.Sprintf("ovirt://%s", id)
+	err = r.client.Update(context.Background(), &node)
+	if err != nil {
+		klog.Errorf("Updating node %s failed %v", node.Name, err)
+		return reconcile.Result{}, fmt.Errorf("Failed updating node %s: %v", node.Name, err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *providerIDReconciler) fetchOvirtVmID(nodeName string) (string, error) {
+	c, err := r.getConnection("openshift-machine-api", "ovirt-credentials")
+	if err != nil {
+		return "", err
+	}
+	send, err := c.SystemService().VmsService().List().Search(fmt.Sprintf("name=%s", nodeName)).Send()
+	if err != nil {
+		r.log.Error(err, "Failed to find VM", "VM name", nodeName)
+		return "", err
+	}
+	vms := send.MustVms().Slice()
+	if len(vms) != 1 {
+		return "", fmt.Errorf("expected to get 1 VM but got %v", len(vms))
+	}
+	return vms[0].MustId(), nil
+}
+
+func Add(mgr manager.Manager, opts manager.Options) error {
+	reconciler, err := NewProviderIDReconciler(mgr)
+
+	if err != nil {
+		return fmt.Errorf("error building reconciler: %v", err)
+	}
+
+	c, err := controller.New("provdierID-controller", mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+
+	//Watch node changes
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func NewProviderIDReconciler(mgr manager.Manager) (*providerIDReconciler, error) {
+	log.SetLogger(klogr.New())
+	r := providerIDReconciler{
+		log:    log.Log.WithName("controllers").WithName("providerID-reconciler"),
+		client: mgr.GetClient(),
+	}
+	r.fetchProviderIDFunc = r.fetchOvirtVmID
+	return &r, nil
+}
+
+func (r *providerIDReconciler) getConnection(namespace, secretName string) (*ovirtsdk.Connection, error) {
+	var err error
+	if r.ovirtApi == nil || r.ovirtApi.Test() != nil {
+		// session expired or some other error, re-login.
+		r.ovirtApi, err = createApiConnection(r.client, namespace, secretName)
+	}
+	return r.ovirtApi, err
+}
+
+//createApiConnection returns a a client to oVirt's API endpoint
+func createApiConnection(client client.Client, namespace string, secretName string) (*ovirtsdk.Connection, error) {
+	creds, err := clients.GetCredentialsSecret(client, namespace, secretName)
+
+	if err != nil {
+		klog.Infof("Failed getting credentials for namespace %s, %s", namespace, err)
```

Comment:
Could pass a logger in as an argument to this method to be able to log using the `providerIDReconciler` logger here